Source/MQTTnet/Internal/AsyncLock.cs

```diff
@@ -7,9 +7,11 @@ namespace MQTTnet.Internal
     // From Stephen Toub (https://blogs.msdn.microsoft.com/pfxteam/2012/02/12/building-async-coordination-primitives-part-6-asynclock/)
     public sealed class AsyncLock : IDisposable
     {
-        readonly SemaphoreSlim _semaphore = new SemaphoreSlim(1, 1);
+        readonly object _syncRoot = new object();
```

Comment:
why lock a semaphore?

Reply:
So make sure that no other thread will access the semaphore while it is being disposed. It will be set to _null_ when the async lock being released but in some cases (race conditions) I got ObjectDisposedExceptions. With the lock the semaphore is either available and ready to use or _null_ (whis is fine for the releaser).